packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualPositionManager.sol

```diff
@@ -474,15 +475,39 @@ contract PerpetualPositionManager is FeePayer, AdministrateeInterface {
         tokenCurrency.burn(numTokens.rawValue);
     }
 
+    /**
+     * @notice Burns `numTokens` of `tokenCurrency` to decrease sponsors position size, without sending back any
+     * `collateralCurrency`. This is done by a sponsor to increase position CR.
+     * @dev Can only be called by token sponsor. This contract must be approved to spend `numTokens` of `tokenCurrency`.
+     * @param numTokens is the number of tokens to be burnt for a commensurate amount of collateral.
+     */
     function repay(FixedPoint.Unsigned memory numTokens)
         public
         notEmergencyShutdown()
         noPendingWithdrawal(msg.sender)
         fees()
         nonReentrant()
-        returns (FixedPoint.Unsigned memory amountRepaired)
     {
-        // TODO: implementation
+        PositionData storage positionData = _getPositionData(msg.sender);
+        require(!numTokens.isGreaterThan(positionData.tokensOutstanding), "Invalid token amount");
+
+        // Decrease the sponsors position tokens size. Ensure it is above the min sponsor size OR all tokens are repaid.
```

Comment:
I think this opens up a subtle edge case where the user could end up with nonzero collateral, but 0 tokens outstanding. This wouldn't _lock_ the collateral, but to avoid the divide-by-zero, we just say the CR of such a position is 0. This means they wouldn't be able to redeem or withdraw immediately. They would have to either create or do a slow withdrawal.

As a side note: unless we choose to change how the contract handles this, we should probably document that a position with collateral but no tokens is not allowed and maybe change our checks to explicitly disallow it.

Reply:
So how about we simply let you repay to minimum sponsor size?

Reply:
@chrismaree Yep, that's how it _should_ work :). I think we do it this way so that the first position is allowed to be created against a GCR of 0/0, which resolves to 0 rather than infinity (which would never allow a position to be created).